aptos-move/mvhashmap/src/versioned_data.rs

```diff
@@ -260,38 +256,58 @@ impl<K: Hash + Clone + Debug + Eq, V: TransactionWrite> VersionedData<K, V> {
             },
-            Occupied(o) => {
-                assert!(
-                    if let EntryCell::Write(i, v, layout) = &o.get().cell {
-                        // base value may have already been provided by another transaction
-                        // executed simultaneously and asking for the same resource.
-                        // Value from storage must be identical, but then delayed field
-                        // identifier exchange could've modiefied it.
-                        //
-                        // If maybe_layout is None, they are required to be identical
-                        // If maybe_layout is Some, there might have been an exchange
-                        // Assert the length of bytes for efficiency (instead of full equality)
-                        layout.is_some() == maybe_layout.is_some()
-                            && *i == 0
-                            && (layout.is_some() || v.bytes().map(|b| b.len()) == bytes_len)
-                    } else {
-                        true
+            Occupied(mut o) => {
+                if let EntryCell::Write(i, existing_value) = &o.get().cell {
+                    assert!(*i == 0);
+                    match (existing_value, &value) {
+                        (RawFromStorage(ev), RawFromStorage(v)) => {
+                            // Base value from storage needs to be identical
+                            // Assert the length of bytes for efficiency (instead of full equality)
+                            assert!(
+                                *i == 0
+                                    && v.bytes().map(|b| b.len()) == ev.bytes().map(|b| b.len())
+                            )
+                        },
+                        (Exchanged(_, _), RawFromStorage(_)) => {
+                            // Stored value contains more info, nothing to do.
+                        },
+                        (RawFromStorage(_), Exchanged(_, _)) => {
+                            // Received more info, update
+                            o.insert(CachePadded::new(Entry::new_write_from(0, value)));
+                        },
+                        (Exchanged(ev, e_layout), Exchanged(v, layout)) => {
+                            // base value may have already been provided by another transaction
+                            // executed simultaneously and asking for the same resource.
+                            // Value from storage must be identical, but then delayed field
+                            // identifier exchange could've modiefied it.
+                            //
+                            // If maybe_layout is None, they are required to be identical
+                            // If maybe_layout is Some, there might have been an exchange
+                            // Assert the length of bytes for efficiency (instead of full equality)
+                            assert!(
```

Comment:
This assert condition is really hard to follow.
why not first assert_eq!(e_layout.is_some(), layout.is_some())
Then if layout.is_none(), assert_eq!(v.bytes().map(|b| b.len()), ev.bytes().map(|b| b.len())));
(or for the latter, assert!(layout.is_some() || bytes equality), like above.

Reply:
Done.